File: deploy/deploy-package/game.py
"""Room game logic (Rust game.rs equivalent).

The Room class holds all per-room state. Game logic methods mutate Room in place.
The caller (ws_handler) is responsible for acquiring room.lock before calling these.
"""

from __future__ import annotations

import asyncio
from enum import Enum

from words import pick_words
from models import ScoreEntry
from db import save_room as db_save_room

DRAW_SECS: int = 60


class GamePhase(str, Enum):
    WAITING = "waiting"
    NEW_ROUND = "new_round"
    DRAWING = "drawing"
    ROUND_RESULT = "round_result"
    GAME_OVER = "game_over"


class Player:
    __slots__ = ("id", "nickname", "score", "has_drawn", "is_connected")

    def __init__(self, id: str, nickname: str) -> None:
        self.id = id
        self.nickname = nickname
        self.score = 0
        self.has_drawn = False
        self.is_connected = True


class Room:
    """Per-room state and game logic. Protected by `lock`."""

    def __init__(self, id: str) -> None:
        self.id = id
        self.owner_id: str = ""  # set when first player joins
        self.players: list[Player] = []
        self.phase: GamePhase = GamePhase.WAITING
        self.current_round: int = 0
        self.total_rounds: int = 0
        self.current_drawer_id: str = ""

        # Phase-scoped data
        self.word: str = ""
        self.word_category: str = ""  # hint for guessers, e.g. "燕云十六声武器"
        self.word_options: list[tuple[str, str]] = []  # (word, category)
        self.guessed_players: dict[str, int] = {}  # player_id -> score earned
        self.guess_counts: dict[str, int] = {}  # player_id -> guesses used this round
        self.seconds_left: int = 0
        self.rankings: list[ScoreEntry] = []
        self.round_scores: list[ScoreEntry] = []

        self.lock = asyncio.Lock()

    # ------------------------------------------------------------------
    # helpers
    # ------------------------------------------------------------------

    def _connected(self) -> list[Player]:
        return [p for p in self.players if p.is_connected]

    def _player_name(self, player_id: str) -> str:
        for p in self.players:
            if p.id == player_id:
                return p.nickname
        return ""

    # ------------------------------------------------------------------
    # public API (matching Rust impl Room)
    # ------------------------------------------------------------------

    def can_start(self) -> tuple[bool, str]:
        n = len(self._connected())
        if n < 2:
            return False, "至少需要2名玩家"
        if self.phase not in (GamePhase.WAITING, GamePhase.GAME_OVER):
            return False, "游戏已在进行中"
        return True, ""

    def start_game(self, rounds: int = 0) -> None:
        if self.phase == GamePhase.GAME_OVER:
            self.phase = GamePhase.WAITING
        for p in self.players:
            p.score = 0
            p.has_drawn = False
        n = len(self._connected())
        self.total_rounds = rounds if rounds > 0 else n
        self.current_round = 1
        self._go_new_round()
        db_save_room(self)

    def select_word(self, player_id: str, idx: int) -> str:
        """Returns the selected word. Raises ValueError on invalid state/input."""
        if self.phase != GamePhase.NEW_ROUND:
            raise ValueError("不在选词阶段")
        if player_id != self.current_drawer_id:
            raise ValueError("只有画手可以选词")
        if idx < 0 or idx >= len(self.word_options):
            raise ValueError("无效选项")
        word, category = self.word_options[idx]
        if not word:
            raise ValueError("无效的词")
        self.word = word
        self.word_category = category
        self.seconds_left = DRAW_SECS
        self.guessed_players.clear()
        self.guess_counts.clear()
        self.phase = GamePhase.DRAWING
        db_save_room(self)
        return word

    def submit_guess(self, player_id: str, text: str) -> int | None:
        """Returns score if correct, None if wrong/duplicate. Raises on invalid state."""
        if self.phase != GamePhase.DRAWING:
            raise ValueError("不在画画阶段")
        if player_id == self.current_drawer_id:
            raise ValueError("画手不能猜词")
        if player_id in self.guessed_players:
            return None
        # Limit to 3 guesses per player per round
        used = self.guess_counts.get(player_id, 0)
        if used >= 3:
            raise ValueError("本轮猜测次数已用完")
        self.guess_counts[player_id] = used + 1
        if text.strip().lower() != self.word.strip().lower():
            return None
        # Correct guess — score by order
        n = len(self.guessed_players)
        score = 30 if n == 0 else (20 if n == 1 else 10)
        self.guessed_players[player_id] = score
        for p in self.players:
            if p.id == player_id:
                p.score += score
                break
        return score

    def tick_timer(self) -> int | None:
        """Decrement timer. Returns new seconds_left, or None if already 0."""
        if self.phase != GamePhase.DRAWING:
            return None
        if self.seconds_left == 0:
            return None
        self.seconds_left -= 1
        return self.seconds_left

    def all_guessed(self) -> bool:
        if self.phase != GamePhase.DRAWING:
            return False
        guesser_count = sum(
            1 for p in self.players
            if p.is_connected and p.id != self.current_drawer_id
        )
        return len(self.guessed_players) >= guesser_count

    def end_round(self) -> tuple[str, list[ScoreEntry]]:
        """Transition to ROUND_RESULT. Returns (answer, scores)."""
        scores = [
            ScoreEntry(
                player_id=pid,
                player_name=self._player_name(pid),
                score=s,
            )
            for pid, s in self.guessed_players.items()
        ]
        self.round_scores = scores
        self.phase = GamePhase.ROUND_RESULT
        db_save_room(self)
        return self.word, scores

    def advance_round(self) -> GamePhase:
        """Mark drawer done, advance to next round or game over."""
        for p in self.players:
            if p.id == self.current_drawer_id:
                p.has_drawn = True
                break

        # Reset has_drawn for next rotation if all connected players have drawn
        if all(p.has_drawn for p in self._connected()):
            for p in self.players:
                p.has_drawn = False

        if self.current_round < self.total_rounds:
            self.current_round += 1
            self._go_new_round()
            return GamePhase.NEW_ROUND

        rankings = sorted(
            (
                ScoreEntry(
                    player_id=p.id,
                    player_name=p.nickname,
                    score=p.score,
                )
                for p in self.players
            ),
            key=lambda e: e.score,
            reverse=True,
        )
        self.rankings = rankings
        self.phase = GamePhase.GAME_OVER
        db_save_room(self)
        return GamePhase.GAME_OVER

    # ------------------------------------------------------------------
    # internal
    # ------------------------------------------------------------------

    def _go_new_round(self) -> None:
        drawer = next(
            (p for p in self.players if p.is_connected and not p.has_drawn),
            None,
        )
        if drawer is None:
            raise RuntimeError("no drawer available")
        self.current_drawer_id = drawer.id
        self.word_options = pick_words(3)
        self.phase = GamePhase.NEW_ROUND
        self.word = ""

File: server/src/game.rs
use crate::types::*;
use crate::words::pick_words;

const DRAW_SECS: u8 = 60;

impl Room {
    pub fn can_start(&self) -> Result<(), &'static str> {
        let n = self.players.iter().filter(|p| p.is_connected).count();
        if n < 2 {
            return Err("至少需要2名玩家");
        }
        if !matches!(self.state, GameState::Waiting | GameState::GameOver { .. }) {
            return Err("游戏已在进行中");
        }
        Ok(())
    }

    pub fn start_game(&mut self) {
        // Reset if coming from GameOver
        if matches!(self.state, GameState::GameOver { .. }) {
            self.state = GameState::Waiting;
        }
        for p in &mut self.players {
            p.score = 0;
            p.has_drawn = false;
        }
        self.total_rounds = self.players.iter().filter(|p| p.is_connected).count() as u8;
        self.current_round = 1;
        self.go_new_round();
    }

    fn go_new_round(&mut self) {
        // Find first connected player who hasn't drawn yet
        let drawer = self
            .players
            .iter()
            .filter(|p| p.is_connected && !p.has_drawn)
            .map(|p| p.id.clone())
            .next()
            .expect("no drawer available");
        self.current_drawer_id = drawer.clone();
        let words = pick_words(3);
        self.state = GameState::NewRound {
            drawer_id: drawer,
            word_options: words,
        };
    }

    pub fn select_word(&mut self, player_id: &str, idx: usize) -> Result<(), &'static str> {
        let word = match &self.state {
            GameState::NewRound {
                drawer_id,
                word_options,
            } => {
                if player_id != drawer_id {
                    return Err("只有画手可以选词");
                }
                word_options.get(idx).ok_or("无效选项")?.clone()
            }
            _ => return Err("不在选词阶段"),
        };
        // Sanity check: can't have 0-length word
        if word.is_empty() {
            return Err("无效的词");
        }

        let drawer_id = self.current_drawer_id.clone();
        self.state = GameState::Drawing {
            drawer_id,
            word,
            seconds_left: DRAW_SECS,
            guessed_players: std::collections::HashMap::new(),
        };
        Ok(())
    }

    pub fn submit_guess(&mut self, player_id: &str, text: &str) -> Result<Option<u32>, &'static str> {
        match &mut self.state {
            GameState::Drawing {
                drawer_id,
                word,
                guessed_players,
                ..
            } => {
                if player_id == drawer_id {
                    return Err("画手不能猜词");
                }
                if guessed_players.contains_key(player_id) {
                    return Ok(None);
                }
                if text.trim().to_lowercase() != word.trim().to_lowercase() {
                    return Ok(None);
                }

                let score: u32 = match guessed_players.len() {
                    0 => 30,
                    1 => 20,
                    _ => 10,
                };
                guessed_players.insert(player_id.to_string(), score);
                if let Some(p) = self.players.iter_mut().find(|p| p.id == player_id) {
                    p.score += score;
                }
                Ok(Some(score))
            }
            _ => Err("不在画画阶段"),
        }
    }

    pub fn tick_timer(&mut self) -> Option<u8> {
        match &mut self.state {
            GameState::Drawing { seconds_left, .. } => {
                if *seconds_left == 0 {
                    return None;
                }
                *seconds_left -= 1;
                Some(*seconds_left)
            }
            _ => None,
        }
    }

    pub fn all_guessed(&self) -> bool {
        match &self.state {
            GameState::Drawing {
                drawer_id,
                guessed_players,
                ..
            } => {
                let guesser_count = self
                    .players
                    .iter()
                    .filter(|p| p.is_connected && p.id != *drawer_id)
                    .count();
                guessed_players.len() >= guesser_count
            }
            _ => false,
        }
    }

    pub fn end_round(&mut self) -> (String, Vec<ScoreEntry>) {
        let (word, guessed) = match &self.state {
            GameState::Drawing {
                word,
                guessed_players,
                ..
            } => (word.clone(), guessed_players.clone()),
            _ => return (String::new(), vec![]),
        };
        let scores: Vec<ScoreEntry> = guessed
            .iter()
            .map(|(id, &score)| {
                let name = self
                    .players
                    .iter()
                    .find(|p| &p.id == id)
                    .map(|p| p.nickname.clone())
                    .unwrap_or_default();
                ScoreEntry {
                    player_id: id.clone(),
                    player_name: name,
                    score,
                }
            })
            .collect();
        self.state = GameState::RoundResult {
            answer: word.clone(),
            scores: scores.clone(),
        };
        (word, scores)
    }

    pub fn advance_round(&mut self) -> Option<GameState> {
        // Mark the current drawer as having drawn
        let drawer_id = self.current_drawer_id.clone();
        if let Some(p) = self.players.iter_mut().find(|p| p.id == drawer_id) {
            p.has_drawn = true;
        }

        // Check if there are more players who haven't drawn
        let more_to_draw = self
            .players
            .iter()
            .filter(|p| p.is_connected)
            .any(|p| !p.has_drawn);

        if more_to_draw {
            self.current_round += 1;
            self.go_new_round();
            None
        } else {
            let mut rankings: Vec<ScoreEntry> = self
                .players
                .iter()
                .map(|p| ScoreEntry {
                    player_id: p.id.clone(),
                    player_name: p.nickname.clone(),
                    score: p.score,
                })
                .collect();
            rankings.sort_by(|a, b| b.score.cmp(&a.score));
            let state = GameState::GameOver { rankings };
            self.state = state.clone();
            Some(state)
        }
    }

    #[allow(dead_code)]
    pub fn restart(&mut self) {
        for p in &mut self.players {
            p.score = 0;
            p.has_drawn = false;
        }
        self.state = GameState::Waiting;
        self.current_round = 0;
        self.total_rounds = 0;
        self.current_drawer_id = String::new();
    }
}
